### ISSUE TITLE
Data loader variability when number of images indivisible by # GPUs

### ISSUE THREAD

**Issue by paulbisso** [NVIDIA/DALI#389]
My validation image set has 150 images on an AWS p2.8xlarge instance, and want to make full use of the 8 GPUs. 

My test set (600 imgs) and training set (2000 imgs) are both divisible by 8, and I have no issues with variability on an epoch-by-epoch basis. In other words, if I evaluate the loss on my test set 25 times in a row, I get the same total loss 25 times (as measured by pytorch's nn.CrossEntropyLoss() function). The loader starts from the first file in the directory and proceeds smoothly to the last file in the directory. Great! 

For my validation directory, however, I can evaluate the loss on my validation set 25 times in a row, with no shuffling of images, and get a wide range of different values. After a bit of digging, I found that the loader is either clipping the set or artificially extending the set. For instance, say I set a per GPU batch size of 2. That means each loop of my enumerated data loader is going through 16 images. Most of the time it will load 144 images and lose 6. Depending on how the sequence of images is set up, though, sometimes it will go through 160 images. I haven't found a way to control this behavior, even with resetting the data loader -- it's because the reset function in pytorch.py looks like: 

```python
if self._counter >= self._size:
            self._counter = self._counter % self._size
``` 

This presents an obvious problem: if my validation set is changing every time, my validation loss changes every time. Which makes it very hard to know how well my model is doing wrt overfitting and when to stop training. So what I'm wondering is -- how can I get my DALI data loader to load through a full epoch every time without clipping and without abridging my data set to be a multiple of 8? Code for my external source iterator, pipeline, DALIGenericIterator and validation checking function are below. 

```python
# Define the NVIDIA DALI iterator that will feed into the data loading pipeline. 
# The 'image_dir' passed into __init__ is assumed to be a top-level image directory. In other words, there should be no image
# files immediately within that directory. The algorithm below will ignore all image files that are directly within the
# top-level image directory. Rather, the top-level directory should contain exclusively other directories. 

# Each of those sub-directories should contain exclusively .JPEG image files.
# Files within each sub-directory are assumed to belong to the same image class. 

class ExternalInputIterator(object):
    def __init__(self, batch_size, image_dir, shuffle_flag):
        self.images_dir = image_dir
        self.files = []
        self.shuffle_flag = shuffle_flag
        self.num_files = 0
        assert isinstance(shuffle_flag, bool)
        # Check which of the entries in the top directory are themselves directories. 
        dirfiles = os.listdir(image_dir)
        num_imgdirs = 0
        self.img_label_dict = dict() # storage denoting which label number corresponds with which label name.
        
        # NOTE: the below will only grab JPG/JPEG files, and would need to be modified to grab other types of image files.
        # NOTE: even if the below were modified to grab other types of image files, the DALI image processing pipeline
        # uses a JPEG decoder to load in the files, so just keep that in mind.
        for file in dirfiles:
            # Grab the name of every entry in the top directory.
            fullname = os.path.join(image_dir, file)
            # If an entry is a directory, grab the filenames for all the JPG/JPEG files in that directory.
            if os.path.isdir(fullname):
                imgfiles = os.listdir(fullname)
                for imgfile in imgfiles:
                    thisimgpath = os.path.join(fullname,imgfile)
                    if os.path.isfile(thisimgpath) and (thisimgpath.lower().endswith(".jpg") or thisimgpath.lower().endswith(".jpeg")):
                        self.img_label_dict[thisimgpath] = num_imgdirs
                        self.files.append(thisimgpath)
                num_imgdirs += 1
        assert num_imgdirs > 0
        # Store the batch size.
        self.batch_size = batch_size
        # Shuffle the input filenames. 
        if self.shuffle_flag:
            shuffle(self.files)

    def __iter__(self):
        self.i = 0
        self.n = len(self.files)
        return self

    def __next__(self):
        batch = []
        labels = []
        for _ in range(self.batch_size):
            jpeg_filename, label = self.files[self.i], self.img_label_dict[self.files[self.i]]
            self.num_files += 1
            f = open(jpeg_filename, 'rb')
            # Later steps in the DALI image processing pipeline (the resize operation, in particular) expect uint8s. 
            batch.append(np.frombuffer(f.read(), dtype = np.uint8))
            labels.append(np.array([label], dtype = np.uint8))
            self.i = (self.i + 1) % self.n
        return (batch, labels)
    
    next = __next__

```

```python
# Define the image processing pipeline for validation.
# If reproducibility is desired, on the input to super(ValidPipeline, self), set seed = some # other than -1. 

class ValidPipeline(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, image_dir):
        super(ValidPipeline, self).__init__(batch_size, num_threads, device_id, seed = 12)
        self.input = ops.ExternalSource()
        self.input_label = ops.ExternalSource()
        self.decode = ops.nvJPEGDecoder(device = "mixed", output_type = types.RGB)
        self.resize = ops.Resize(device="gpu")
        self.resize_rng = ops.Uniform(range = (224,224))
        self.crop = ops.Crop(device="gpu", crop=(224,224))
        self.norm = ops.NormalizePermute(device = "gpu",
                                        height = 224,
                                        width = 224,
                                        mean = [123.675, 116.28, 103.53],
                                        std = [58.395, 57.12, 57.375])

    def define_graph(self):
        self.jpegs = self.input()
        self.labels = self.input_label()
        images = self.decode(self.jpegs)
        images = self.resize(images, resize_shorter=self.resize_rng())
        images = self.crop(images)
        images = self.norm(images)
        return (images, self.labels)
    
    def iter_setup(self):
        (images, labels) = iter_valid.next()
        self.feed_input(self.jpegs, images)
        self.feed_input(self.labels, labels)
```

```python
N = 8
valid_batch = 2
num_valid_images = 150
valid_dir = 'data/valid'
eii_valid = ExternalInputIterator(batch_size=valid_batch, image_dir=valid_dir, shuffle_flag=False)
iter_valid = iter(eii_valid)
# Validation loader.
pipe_v =  [ValidPipeline, (0, 2)]
pipe_name, label_range = pipe_v
print ("RUN: "  + pipe_name.__name__)        
pipes = [pipe_name(batch_size=valid_batch, num_threads=2, device_id = device_id, image_dir = valid_dir) for device_id in range(N)]

valid_loader = DALIGenericIterator(pipes, ['data', 'label'], num_valid_images)
```

```python

def valid_check(model,loaders,criterion,save_path,load_model):
    if load_model:
        model.load_state_dict(torch.load(save_path))
    model.eval()
    valid_loss = 0

    if use_dali:
        for batch_idx, batch_data in enumerate(loaders['valid']):
            # stack data and move targets to GPU
            data = torch.stack([image for gpu_batch in batch_data for image in gpu_batch[0][0]])
            target = torch.stack([target.to(torch.long) for gpu_batch in batch_data for target in gpu_batch[1][0]])
            target = target.to(device)

            ## update the average validation loss
            output = model.forward(data)
            loss = criterion(output,target)
            valid_loss += 1/(batch_idx+1)*(loss.data-valid_loss)
       loaders['valid'].reset()

    # print validation statistics
    if load_model:
        print('Loaded model validation loss: {:.6f}'.format(
        valid_loss
        ))

    return valid_loss
```

Running the following 5 times: 
```python
best_valid_loss = valid_check(model2, loader_dict, criterion,'model_take1.pt', True)
```

yields the following:
Loaded model validation loss: 0.726238
Loaded model validation loss: 0.696305
Loaded model validation loss: 0.858156
Loaded model validation loss: 0.717103
Loaded model validation loss: 0.730114

**Comment by ptrendx** [comment]
You actually get 10 duplicate images and not 6 less images every time (it goes over instead of clipping). I believe the native pyTorch data loader clips.

I'm not a pyTorch expert, so I can't tell you exactly how to solve it in pyTorch, but I can tell you how we solved this problem (we want to validate exactly on validation dataset and not clip/overflow) in MXNet plugin. Basically MXNet's `DataBatch` has a `pad` parameter that would tell you how many images are "true" images and how many are just padding. DALI itself does not have a notion of padding batches that are not full, but letting it overflow and just marking the overflow images are pad still works (you will get different ordering of images for validation each time, but it does not matter as you will always get the full dataset). The option to enable that in the MXNet plugin is `fill_last_batch=False`. You can look at that plugin here: https://github.com/NVIDIA/DALI/blob/master/dali/python/nvidia/dali/plugin/mxnet.py#L211-L224 for info how to compute the pad value. Then I guess on the pyTorch side if you see `pad > 0` you could push only portion of the batch to the model instead of the entire batch. Does that make sense?

**Comment by ptrendx** [comment]
Adding @JanuszL for visibility - actually I think we should be able to do something like what I described in the previous post to enable that option in pyTorch plugin.
@paulbisso if you test that method and it works, please consider making a PR.

**Comment by paulbisso** [comment]
@ptrendx I think this method will work. I drew up a modification to pytorch.py independently that is working at getting the exact right number of samples per epoch, but the order changes every time.

Doing it with padding as you suggest seems like it would get both the number of samples ordering right. 

I'll test and initiate a pull request shortly. Thanks!

**Comment by paulbisso** [comment]
@ptrendx I've drawn something up that gets the epoch size right each time; it's similar to what you did for MXNet, but without any padding option. It gets the right number of images, but different ordering. It's too bad there is no concept of ordering internal to DALI, as this would reduce the number of downstream code modifications needed to plug DALI in. 

For instance, I'm using the [CrossEntropyLoss()](https://pytorch.org/docs/stable/nn.html#torch.nn.CrossEntropyLoss) function for loss in my model -- it combines a softmax with a negative log likelihood loss. When the images are ordered identically, there is no clipping / overage and the batch size is kept the same, as in multiple sequential runs of a validation loss check using the pytorch DataLoader, then CrossEntropyLoss() obviously returns the same loss value for a given model and set of data. However, if either the ordering or the batch size changes, the total loss will vary from run to run. This is due to the 'reduction' parameter in CrossEntropyLoss(). Its default is set to 'mean'. The averaging process, done for the user's convenience, will produce varying loss values on identical data / models if the ordering of the data changes. To avoid this, the user needs to set 'reduction='sum'' and then average over the whole epoch. That will eliminate the impact of ordering.

i.e.

```python
loss = nn.CrossEntropyLoss(reduction='sum')
```

This is worth a mention in the docs, as it wasn't immediately obvious why the loss disparity between the pytorch DataLoader and my edited DALI pytorch.py was occurring as I ran through my tests.

**Comment by ptrendx** [comment]
Ok, I see. The `mean` option is not exactly a mean then ;-). 
Making ordering work the same would require being able to actually reset a data loader, which is slightly tricky because of prefetching, but we should consider implementing this. 
Anyway, thank you @paulbisso  for investigating that - will you make a PR for the option in pyTorch plugin to give not full batches? 
@JanuszL - I agree we should find a place to mention this behavior in the docs.

**Comment by paulbisso** [comment]
@ptrendx Done, pull request #391 made.
